org.eclipse.xtext.xbase/src/org/eclipse/xtext/xtype/util/XFunctionTypeRefs.java

```diff
@@ -19,25 +19,29 @@
 import org.eclipse.xtext.common.types.JvmVoid;
 import org.eclipse.xtext.common.types.TypesFactory;
 import org.eclipse.xtext.common.types.access.impl.ClassURIHelper;
+import org.eclipse.xtext.common.types.util.Primitives;
 import org.eclipse.xtext.xbase.lib.Functions;
 import org.eclipse.xtext.xbase.lib.Procedures;
 
 /**
  * @author Anton Kosyakov - Initial contribution and API
  */
 public class XFunctionTypeRefs {
+	
+	private static final Primitives primitives = new Primitives();
+
+	private static final ClassURIHelper classURIHelper = new ClassURIHelper();
+	
 	public static URI computeTypeUri(boolean procedure, int functionParamCount) {
 		return URI.createURI(buildUri(procedure, functionParamCount));
 	}
 
 	public static String buildUri(boolean procedure, int functionParamCount) {
 		int paramCount = Math.min(6, functionParamCount);
 		if (procedure) {
-			return "java:/Objects/" + Procedures.class.getCanonicalName() + "#" + Procedures.class.getCanonicalName()
-					+ "$Procedure" + Integer.valueOf(paramCount);
+			return classURIHelper.getFullURI(Procedures.class).toString() + "$Procedure" + paramCount;
 		}
-		return "java:/Objects/" + Functions.class.getCanonicalName() + "#" + Functions.class.getCanonicalName()
-				+ "$Function" + Integer.valueOf(paramCount);
+		return classURIHelper.getFullURI(Functions.class).toString() + "$Function" + paramCount;
```

Comment:
I'm not sure that converting this to a URI and subsequently calling toString again is a good tradeoff. 
You may want to replace the literal `"java:/Objects/"` by a concatenation from URIHelperConstants, though.

Since `buildURI` is only used from within this class, it could also be an option to overhaul the flow and do something based on Class.forName instead or even keep an immutable list of the 2x6 allowed combinations such that we can return the URI in constant time without any string / URI manipulation.

Reply:
Could you please elaborate on the second point above?

Reply:
> even keep an immutable list of the 2x6 allowed combinations such that we can return the URI in constant time without any string / URI manipulation.

@szarnekow Updated the code. The method `buildUri` returns a URI in constant time.